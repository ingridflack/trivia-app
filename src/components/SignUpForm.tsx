import {
  Avatar,
  AvatarGroup,
  Button,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Input,
} from "@chakra-ui/react";
import { Controller, SubmitHandler, useForm } from "react-hook-form";
import * as AuthService from "../services/authService";
import { useNavigate } from "react-router-dom";
import { useToast } from "@chakra-ui/react";
import { AVATAR_OPTIONS } from "../constants/user";

interface LoginFormValues {
  email: string;
  password: string;
  name: string;
  username: string;
  avatar?: string;
  passwordConfirmation: string;
}

export default function SignUpForm() {
  const {
    handleSubmit,
    register,
    control,
    formState: { errors, isSubmitting },
  } = useForm<LoginFormValues>();
  const navigate = useNavigate();
  const toast = useToast();

  const onSubmit: SubmitHandler<LoginFormValues> = async (values) => {
    try {
      await AuthService.register({
        ...values,
      });

      toast({
        title: "Account created",
        description: "Account created successfully. Please log in.",
        status: "success",
        duration: 5000,
        isClosable: true,
        position: "top-right",
      });

      navigate("/login");
    } catch (error) {
      console.log(error);

      toast({
        title: "Oops!",
        description: "Something went wrong. Try again.",
        status: "error",
        duration: 5000,
        isClosable: true,
        position: "top-right",
      });
    }
  };

  return (
    <form
      onSubmit={handleSubmit(onSubmit)}
      style={{
        width: "100%",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
      }}
    >
      <FormControl isInvalid={!!errors.email}>
        <FormLabel htmlFor="email" fontSize="small" marginBottom="4px">
          Email address
        </FormLabel>
        <Input
          type="email"
          placeholder="Email"
          id="email"
          {...register("email", {
            required: "Email is required.",
            pattern: {
              value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
              message: "Invalid email address",
            },
          })}
        />
        <FormErrorMessage>
          {errors.email && errors.email.message}
        </FormErrorMessage>
      </FormControl>

      <FormControl isInvalid={!!errors.name}>
        <FormLabel
          marginTop="16px"
          htmlFor="name"
          fontSize="small"
          marginBottom="4px"
        >
          Name
        </FormLabel>
        <Input
          type="text"
          placeholder="Name"
          id="name"
          {...register("name", {
            required: "Name is required.",
          })}
        />
        <FormErrorMessage>
          {errors.name && errors.name.message}
        </FormErrorMessage>
      </FormControl>

      <FormControl isInvalid={!!errors.username}>
        <FormLabel
          marginTop="16px"
          htmlFor="username"
          fontSize="small"
          marginBottom="4px"
        >
          Username
        </FormLabel>
        <Input
          type="text"
          placeholder="Username"
          id="username"
          {...register("username", {
            required: "Username is required.",
            pattern: {
              value: /^\S*$/,
              message: "Username should not contain spaces",
            },
          })}
        />
        <FormErrorMessage>
          {errors.name && errors.name.message}
        </FormErrorMessage>
      </FormControl>

      <FormControl isInvalid={!!errors.avatar}>
        <FormLabel
          htmlFor="avatar"
          marginTop="16px"
          fontSize="small"
          marginBottom="4px"
        >
          Avatar
        </FormLabel>
        <AvatarGroup spacing="4px">
          {AVATAR_OPTIONS.map((imageSrc) => (
            <Controller
              key={imageSrc}
              control={control}
              name="avatar"
              render={({ field: { onChange, value, ref } }) => (
                <Avatar
                  src={imageSrc}
                  width="60px"
                  height="60px"
                  borderColor="white"
                  borderWidth={3}
                  bg="purple.500"
                  padding={2}
                  cursor="pointer"
                  onClick={() => onChange(imageSrc)}
                  ref={ref}
                  opacity={value === imageSrc ? 1 : 0.5}
                />
              )}
            />
          ))}
        </AvatarGroup>
        <FormErrorMessage>
          {errors.avatar && errors.avatar.message}
        </FormErrorMessage>
      </FormControl>

      <FormControl isInvalid={!!errors.password}>
        <FormLabel
          marginTop="16px"
          htmlFor="password"
          fontSize="small"
          marginBottom="4px"
        >
          Password
        </FormLabel>
        <Input
          type="password"
          placeholder="Password"
          id="password"
          {...register("password", {
            required: "Password is required.",
            minLength: {
              value: 8,
              message: "Password must be at least 8 characters long",
            },
          })}
        />
        <FormErrorMessage>
          {errors.password && errors.password.message}
        </FormErrorMessage>
      </FormControl>

      <FormControl isInvalid={!!errors.passwordConfirmation}>
        <FormLabel
          marginTop="16px"
          htmlFor="passwordConfirmation"
          fontSize="small"
          marginBottom="4px"
        >
          Password confirmation
        </FormLabel>
        <Input
          type="password"
          placeholder="Password Confirmation"
          id="passwordConfirmation"
          {...register("passwordConfirmation", {
            required: "Password confirmation is required.",
            minLength: {
              value: 8,
              message: "Password must be at least 8 characters long",
            },
            validate: (value) => {
              return value === "password" ? true : "The passwords do not match";
            },
          })}
        />
        <FormErrorMessage>
          {errors.password && errors.password.message}
        </FormErrorMessage>
      </FormControl>

      <Button
        colorScheme="purple"
        size="lg"
        width="100%"
        marginTop="16px"
        isLoading={isSubmitting}
        type="submit"
      >
        Sign Up
      </Button>
    </form>
  );
}
