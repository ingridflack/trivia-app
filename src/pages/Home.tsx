import { Box, Button, Container, Image, Text } from "@chakra-ui/react";
import Header from "../components/Header";
import { Link } from "react-router-dom";

export default function Home() {
  return (
    <>
      <Header />
      <Box width="100%" backgroundColor="#AE79F8" overflow="hidden">
        <Container paddingY="80px" position="relative">
          <Text
            fontSize="6xl"
            lineHeight="1.2"
            fontWeight="bold"
            color="white"
            maxWidth="550"
          >
            Challenge yourself in a trivia match!
          </Text>

          <Button
            backgroundColor="green.400"
            color="white"
            width="180px"
            _hover="green.400"
            marginTop="38px"
            as={Link}
            to="/trivia/config"
          >
            Start a trivia
          </Button>
          <Image
            src="/src/assets/home.svg"
            alt="Home image"
            position="absolute"
            right="-160px"
            bottom="-60px"
            width="674px"
            height="418px"
            pointerEvents="none"
          />
        </Container>
      </Box>
    </>
  );
}
