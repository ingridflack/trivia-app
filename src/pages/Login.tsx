import {
  Box,
  Card,
  Container,
  Image,
  Link,
  Text,
  useToast,
} from "@chakra-ui/react";
import LoginForm from "../components/LoginForm";
import Header from "../components/Header";
import { useNavigate } from "react-router-dom";
import { SubmitHandler, useForm } from "react-hook-form";
import * as AuthService from "../services/authService";
import useAuth from "../hooks/useAuth";

export interface LoginFormValues {
  email: string;
  password: string;
}

export default function Login() {
  const {
    handleSubmit,
    register,
    formState: { errors, isSubmitting },
  } = useForm<LoginFormValues>();
  const navigate = useNavigate();
  const toast = useToast();
  const { saveUserData } = useAuth();

  const onSubmit: SubmitHandler<LoginFormValues> = async (values) => {
    try {
      const { data } = await AuthService.login(values);

      saveUserData(data);

      navigate("/");
    } catch (error) {
      console.log(error);
      toast({
        title: "An error occurred.",
        description: "Unable to login.",
        status: "error",
        duration: 2000,
        isClosable: true,
        position: "top-right",
      });
    }
  };

  return (
    <>
      <Header />

      <Container
        display="flex"
        alignItems="center"
        justifyContent="center"
        minH="calc(var(--chakra-vh) - 50px)" // 100vh - 50px (height of the header)
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          padding="70px"
          shadow="xl"
        >
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="32px" color="gray.700">
              Sign in
            </Text>
            <LoginForm
              onSubmit={handleSubmit(onSubmit)}
              errors={errors}
              isSubmitting={isSubmitting}
              register={register}
            />

            <Box marginTop="20px" display="flex" justifyContent="space-between">
              <Link
                fontSize="small"
                color="gray.500"
                href="/trivia/recovery-email"
              >
                Forgot your password?
              </Link>

              <Link fontSize="small" color="gray.500" href="/sign-up">
                Create an account
              </Link>
            </Box>
          </Box>

          <Image
            src="src/assets/login.svg"
            alt="Login image"
            width="325px"
            height="324px"
            pointerEvents="none"
          />
        </Card>
      </Container>
    </>
  );
}
