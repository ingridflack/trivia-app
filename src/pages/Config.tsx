import { Box, Card, Container, Image, Text } from "@chakra-ui/react";
import TriviaConfig from "../components/TriviaConfig";
import Header from "../components/Header";
import { Footer } from "../components/Footer";

export default function Config() {
  return (
    <>
      <Header />

      <Container
        display="flex"
        alignItems="center"
        justifyContent="center"
        minH="calc(var(--chakra-vh) - 50px)" // 100vh - 50px (height of the header)
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          paddingY="50px"
          paddingX="70px"
          padding={{ base: "20px", md: "50px 70px" }}
          shadow="xl"
        >
          <Image
            src="/src/assets/setup.svg"
            alt="Login image"
            width="388px"
            height="307px"
            pointerEvents="none"
            display={{ base: "none", md: "none", lg: "block" }}
          />
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="20px" color="gray.700">
              Trivia setup
            </Text>
            <TriviaConfig />
          </Box>
        </Card>
      </Container>
      <Footer />
    </>
  );
}
