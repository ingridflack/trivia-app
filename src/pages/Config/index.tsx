import { Box, Card, Container, Image, Text, useToast } from "@chakra-ui/react";
import TriviaConfig from "../../components/TriviaConfig";
import Header from "../../components/Header";
import { Footer } from "../../components/Footer";
import * as UserService from "../../services/userService";
import * as TriviaService from "../../services/triviaService";
import { SubmitHandler, useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";
import { useState } from "react";
import { MultiValue } from "react-select";
import { UserSearchSelectOption } from "../../types/sharedTypes";
import useAuth from "../../hooks/useAuth";

export interface GameConfigValues {
  category: string;
  difficulty: string;
  amount: number;
  gameMode: string;
  type: string;
}

export default function Config() {
  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<GameConfigValues>({
    defaultValues: {
      difficulty: "any",
      type: "any",
      amount: 5,
    },
  });

  const [gameMode, setGameMode] = useState("single");
  const navigate = useNavigate();
  const toast = useToast();
  const [invitedUsers, setInvitedUsers] = useState<
    MultiValue<UserSearchSelectOption>
  >([]);
  const { userData } = useAuth();

  const onSubmit: SubmitHandler<GameConfigValues> = async (values) => {
    try {
      const invitedUsersIds =
        gameMode === "multi"
          ? invitedUsers.map((option) => option.value)
          : undefined;

      const body = {
        ...values,
        invitedUsers: invitedUsersIds,
      };

      const { data } = await TriviaService.createTrivia(body);

      navigate(`/trivia/${data.triviaId}`);
    } catch {
      console.log("error");

      toast({
        title: "Oops!",
        description: "Something went wrong. Try again.",
        status: "error",
        duration: 5000,
        isClosable: true,
        position: "top-right",
      });
    }
  };

  const handleGameModeChange = (value: string) => {
    setGameMode(value);
  };

  const handleInviteUser = (value: MultiValue<UserSearchSelectOption>) => {
    setInvitedUsers(value);
  };

  const handleLoadUsers = async (username: string) => {
    const currentUserId = userData ? userData.id : "";

    const { data } = await UserService.search({ username });
    return data
      .filter((user) => user._id !== currentUserId)
      .map((user) => ({ value: user._id, label: user.username }));
  };

  return (
    <>
      <Header />

      <Container
        as="main"
        display="flex"
        alignItems="center"
        justifyContent="center"
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          paddingY="50px"
          paddingX="70px"
          padding={{ base: "20px", md: "50px 70px" }}
          shadow="xl"
        >
          <Image
            src="/assets/setup.svg"
            alt="Login image"
            width="388px"
            height="307px"
            pointerEvents="none"
            display={{ base: "none", md: "none", lg: "block" }}
          />
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="20px" color="gray.700">
              Trivia setup
            </Text>
            <TriviaConfig
              onLoadUsers={handleLoadUsers}
              onInviteUser={handleInviteUser}
              onGameModeChange={handleGameModeChange}
              onSubmit={handleSubmit(onSubmit)}
              errors={errors}
              register={register}
              isSubmitting={isSubmitting}
              gameMode={gameMode}
              invitedUsers={invitedUsers}
            />
          </Box>
        </Card>
      </Container>
      <Footer />
    </>
  );
}
