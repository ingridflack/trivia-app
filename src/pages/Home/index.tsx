import {
  Box,
  Button,
  Container,
  Heading,
  Image,
  Skeleton,
  Stack,
  Text,
} from "@chakra-ui/react";
import Header from "../../components/Header";
import { Link, useNavigate } from "react-router-dom";
import { PendingTriviaList } from "../../components/PendingTriviaList";
import { useEffect, useState } from "react";
import * as TriviaService from "../../services/triviaService";
import { CreateTriviaBody, PendingTrivia } from "../../types/sharedTypes";
import { Footer } from "../../components/Footer";
import useAuth from "../../hooks/useAuth";
import { TRIVIA_CATEGORIES } from "../../constants/trivia";
import { CategoryCard } from "../../components/CategoryCard";

export default function Home() {
  const [pendingTrivia, setPendingTrivia] = useState<PendingTrivia[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const { userData } = useAuth();
  const navigate = useNavigate();

  useEffect(() => {
    const fetchPendingTrivia = async () => {
      try {
        const { data } = await TriviaService.getPendingTrivia();
        setPendingTrivia(data.pendingTrivia);
      } catch (error) {
        console.log(error);
      } finally {
        setIsLoading(false);
      }
    };

    fetchPendingTrivia();
  }, []);

  const handleCategoryCardClick = async (params: CreateTriviaBody) => {
    try {
      if (!userData) {
        localStorage.setItem("triviaConfig", JSON.stringify(params));
        navigate("/login");
        return;
      }

      const { data } = await TriviaService.createTrivia(params);

      navigate(`/trivia/${data.triviaId}`);
    } catch (error) {
      console.log(error);
    }
  };

  return (
    <>
      <Header />

      <Box as="main">
        <Box width="100%" backgroundColor="purple.500" overflow="hidden">
          <Container
            padding={{ base: "40px 20px", md: "80px 0" }}
            position="relative"
          >
            <Text
              fontSize={{ base: "3xl", md: "6xl" }}
              lineHeight="1.2"
              fontWeight="bold"
              color="white"
              maxWidth="550"
            >
              Challenge yourself in a trivia match!
            </Text>

            <Button
              backgroundColor="green.400"
              color="white"
              width="180px"
              _hover="green.400"
              marginTop="38px"
              as={Link}
              to="/trivia/config"
            >
              Start a trivia
            </Button>
            <Image
              src="/assets/home.svg"
              alt="Home image"
              position="absolute"
              right="-160px"
              bottom="-60px"
              width="674px"
              height="418px"
              pointerEvents="none"
              display={{ base: "none", md: "none", lg: "block" }}
            />
          </Container>
        </Box>

        {userData && (
          <Container padding={{ base: "40px 20px", md: "80px 0 120px" }}>
            <Box as="header" marginBottom={{ base: "20px", md: "40px" }}>
              <Heading fontSize="3xl">Pending trivias</Heading>
            </Box>

            {isLoading ? (
              <Stack
                display="grid"
                gap="16px"
                gridTemplateColumns={{
                  base: "repeat(1fr)",
                  md: "repeat(4, 1fr)",
                }}
              >
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
              </Stack>
            ) : pendingTrivia.length ? (
              <PendingTriviaList pendingTriviaList={pendingTrivia} />
            ) : (
              <Text>No pending trivias</Text>
            )}
          </Container>
        )}

        <Box
          as="section"
          padding={{ base: "40px 20px", md: "80px 0 120px" }}
          backgroundColor="gray.100"
          boxShadow="0px 30px 100px 0 rgba(39, 39, 39, 0.2);"
        >
          <Box
            as="header"
            textAlign="center"
            marginBottom={{ base: "20px", md: "60px" }}
          >
            <Heading fontSize={{ base: "3xl", md: "5xl" }} marginBottom="20px">
              Show off your knowledge
            </Heading>
            <Text fontSize={{ base: "md", md: "xl" }}>
              Take a quiz and impress your friends with what you know!
            </Text>
          </Box>

          <Container
            display="grid"
            gridTemplateColumns={{ base: "repeat(1fr)", md: "repeat(5, 1fr)" }}
            gap="16px"
            padding={{ base: "0", md: "0 0" }}
          >
            {Object.values(TRIVIA_CATEGORIES).map((category) => (
              <CategoryCard
                key={category.title}
                title={category.title}
                icon={category.icon}
                color={category.color}
                onCardClick={() => handleCategoryCardClick(category.params)}
              />
            ))}
          </Container>
        </Box>
      </Box>

      <Footer />
    </>
  );
}
