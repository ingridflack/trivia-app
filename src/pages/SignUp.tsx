import { Box, Card, Container, Text } from "@chakra-ui/react";
import SignUpForm from "../components/SignUpForm";
import Header from "../components/Header";

export default function SignUp() {
  return (
    <>
      <Header />

      <Container
        display="flex"
        alignItems="center"
        justifyContent="center"
        paddingY="32px"
        minH="calc(var(--chakra-vh) - 50px)" // 100vh - 50px (height of the header)
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          padding="32px"
          shadow="xl"
          maxW="635px"
        >
          <Box width="100%">
            <Text fontSize="xx-large" marginBottom="32px" color="gray.700">
              Sign up
            </Text>
            <SignUpForm />
          </Box>
        </Card>
      </Container>
    </>
  );
}
