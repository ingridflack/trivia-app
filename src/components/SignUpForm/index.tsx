import {
  Avatar,
  AvatarGroup,
  Button,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Input,
  Stack,
} from "@chakra-ui/react";
import {
  Control,
  Controller,
  FieldErrors,
  UseFormRegister,
} from "react-hook-form";

import { AVATAR_OPTIONS } from "../../constants/user";
import { checkPasswordConfirmation } from "../../helpers/form";
import { SignUpValues } from "../../pages/SignUp";

export interface SignUpFormProps {
  onSubmit: (e?: React.BaseSyntheticEvent) => Promise<void>;
  errors: FieldErrors<SignUpValues>;
  register: UseFormRegister<SignUpValues>;
  isSubmitting: boolean;
  control: Control<SignUpValues, unknown>;
}

export default function SignUpForm({
  onSubmit,
  errors,
  register,
  isSubmitting,
  control,
}: SignUpFormProps) {
  return (
    <Stack as="form" onSubmit={onSubmit} gap="16px" role="form">
      <FormControl isInvalid={!!errors.email}>
        <FormLabel htmlFor="email">Email address</FormLabel>
        <Input
          type="email"
          placeholder="Email"
          id="email"
          {...register("email", {
            required: "Email is required.",
            pattern: {
              value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
              message: "Invalid email address",
            },
          })}
        />
        <FormErrorMessage>
          {errors.email && errors.email.message}
        </FormErrorMessage>
      </FormControl>

      <Stack direction={{ base: "column", md: "row" }} gap="16px">
        <FormControl isInvalid={!!errors.name}>
          <FormLabel htmlFor="name">Name</FormLabel>
          <Input
            type="text"
            placeholder="Name"
            id="name"
            {...register("name", {
              required: "Name is required.",
            })}
          />
          <FormErrorMessage>
            {errors.name && errors.name.message}
          </FormErrorMessage>
        </FormControl>

        <FormControl isInvalid={!!errors.username}>
          <FormLabel htmlFor="username">Username</FormLabel>
          <Input
            type="text"
            placeholder="Username"
            id="username"
            {...register("username", {
              required: "Username is required.",
              pattern: {
                value: /^\S*$/,
                message: "Username should not contain spaces",
              },
            })}
          />
          <FormErrorMessage>
            {errors.username && errors.username.message}
          </FormErrorMessage>
        </FormControl>
      </Stack>

      <Stack direction={{ base: "column", md: "row" }} gap="16px">
        <FormControl isInvalid={!!errors.password}>
          <FormLabel htmlFor="password">Password</FormLabel>
          <Input
            type="password"
            placeholder="Password"
            id="password"
            {...register("password", {
              required: "Password is required.",
              minLength: {
                value: 8,
                message: "Password must be at least 8 characters long",
              },
            })}
          />
          <FormErrorMessage>
            {errors.password && errors.password.message}
          </FormErrorMessage>
        </FormControl>

        <FormControl isInvalid={!!errors.passwordConfirmation}>
          <FormLabel htmlFor="passwordConfirmation">
            Password confirmation
          </FormLabel>
          <Input
            type="password"
            placeholder="Password Confirmation"
            id="passwordConfirmation"
            {...register("passwordConfirmation", {
              required: "Password confirmation is required.",
              minLength: {
                value: 8,
                message: "Password must be at least 8 characters long",
              },
              validate: { isEqual: checkPasswordConfirmation },
            })}
          />
          <FormErrorMessage>
            {errors.passwordConfirmation && errors.passwordConfirmation.message}
          </FormErrorMessage>
        </FormControl>
      </Stack>

      <FormControl isInvalid={!!errors.avatar}>
        <FormLabel htmlFor="avatar">Avatar</FormLabel>
        <AvatarGroup spacing="4px">
          {AVATAR_OPTIONS.map((imageSrc) => (
            <Controller
              key={imageSrc}
              control={control}
              name="avatar"
              render={({ field: { onChange, value, ref } }) => (
                <Avatar
                  src={imageSrc}
                  width="50px"
                  height="50px"
                  borderColor="white"
                  borderWidth={3}
                  bg="purple.500"
                  padding={2}
                  cursor="pointer"
                  onClick={() => onChange(imageSrc)}
                  ref={ref}
                  opacity={value === imageSrc ? 1 : 0.5}
                  transition={"opacity 0.3s"}
                  _hover={{
                    opacity: 1,
                  }}
                />
              )}
            />
          ))}
        </AvatarGroup>
        <FormErrorMessage>
          {errors.avatar && errors.avatar.message}
        </FormErrorMessage>
      </FormControl>

      <Button
        colorScheme="purple"
        size="lg"
        width="100%"
        isLoading={isSubmitting}
        type="submit"
      >
        Sign Up
      </Button>
    </Stack>
  );
}
