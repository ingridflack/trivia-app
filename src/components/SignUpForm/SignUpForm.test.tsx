import { afterEach, describe, expect, it, vi } from "vitest";
import SignUpForm, { SignUpFormProps } from ".";
import { cleanup, fireEvent, render, renderHook } from "@testing-library/react";

import { FieldErrors, useForm } from "react-hook-form";
import { SignUpValues } from "../../pages/SignUp";

const { result } = renderHook(() => useForm());

const setup = (props: Partial<SignUpFormProps> = {}) =>
  render(
    <SignUpForm
      onSubmit={vi.fn()}
      errors={{}}
      register={vi.fn()}
      isSubmitting={false}
      // @ts-expect-error- ignore the error for the purpose of the test
      control={result.current.control}
      {...props}
    />
  );

const formFieldsData = [
  {
    errors: { email: { message: "Invalid email address" } },
    name: "email",
    regex: /invalid email address/i,
  },
  {
    errors: { name: { message: "Name is required" } },
    name: "name",
    regex: /name is required/i,
  },
  {
    errors: { username: { message: "username is required" } },
    name: "username",
    regex: /username is required/i,
  },
  {
    errors: { password: { message: "Invalid password" } },
    name: "password",
    regex: /invalid password/i,
  },
  {
    errors: { avatar: { message: "Invalid avatar" } },
    name: "avatar",
    regex: /invalid avatar/i,
  },
  {
    errors: {
      passwordConfirmation: { message: "Invalid password confirmation" },
    },
    name: "passwordConfirmation",
    regex: /invalid password confirmation/i,
  },
];

describe("SignUpForm", () => {
  afterEach(() => {
    cleanup();
  });

  it("should match the snapshot", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("should call onSubmit when the form is submitted", () => {
    const mockOnSubmit = vi.fn();
    const { getByRole } = setup({ onSubmit: mockOnSubmit });

    fireEvent.submit(getByRole("form"));
    expect(mockOnSubmit).toHaveBeenCalled();
  });

  it("should render the button in loading state when isSubmitting is true", () => {
    const { getByRole } = setup({ isSubmitting: true });
    const button = getByRole("button", { name: /sign Up/i });
    expect(button).toHaveProperty("disabled", true);
  });

  formFieldsData.forEach(({ errors, name, regex }) => {
    it(`should display the ${name} error message when it's invalid`, () => {
      const props = { errors: errors as FieldErrors<SignUpValues> };
      const { getByText } = setup(props);
      expect(getByText(regex)).toBeTruthy();
    });
  });
});
