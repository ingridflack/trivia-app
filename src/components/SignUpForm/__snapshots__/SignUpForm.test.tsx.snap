// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignUpForm > should match the snapshot 1`] = `
<div>
  <form
    class="chakra-stack css-yd8sa2"
    role="form"
  >
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="email"
        id="field-:r0:-label"
      >
        Email address
      </label>
      <input
        class="chakra-input css-0"
        id="email"
        placeholder="Email"
        type="email"
      />
    </div>
    <div
      class="chakra-stack css-lgk7oi"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="name"
          id="field-:r1:-label"
        >
          Name
        </label>
        <input
          class="chakra-input css-0"
          id="name"
          placeholder="Name"
          type="text"
        />
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="username"
          id="field-:r2:-label"
        >
          Username
        </label>
        <input
          class="chakra-input css-0"
          id="username"
          placeholder="Username"
          type="text"
        />
      </div>
    </div>
    <div
      class="chakra-stack css-lgk7oi"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="password"
          id="field-:r3:-label"
        >
          Password
        </label>
        <input
          class="chakra-input css-0"
          id="password"
          placeholder="Password"
          type="password"
        />
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="passwordConfirmation"
          id="field-:r4:-label"
        >
          Password confirmation
        </label>
        <input
          class="chakra-input css-0"
          id="passwordConfirmation"
          placeholder="Password Confirmation"
          type="password"
        />
      </div>
    </div>
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="avatar"
        id="field-:r5:-label"
      >
        Avatar
      </label>
      <div
        class="chakra-avatar__group css-1rs5v3u"
        role="group"
      >
        <span
          class="chakra-avatar css-1ggfo5f"
        >
          <svg
            aria-label=" avatar"
            class="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          class="chakra-avatar css-1ggfo5f"
        >
          <svg
            aria-label=" avatar"
            class="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          class="chakra-avatar css-1ggfo5f"
        >
          <svg
            aria-label=" avatar"
            class="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          class="chakra-avatar css-1ggfo5f"
        >
          <svg
            aria-label=" avatar"
            class="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          class="chakra-avatar css-1ggfo5f"
        >
          <svg
            aria-label=" avatar"
            class="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          class="chakra-avatar css-1ggfo5f"
        >
          <svg
            aria-label=" avatar"
            class="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
    </div>
    <button
      class="chakra-button css-uxt1e8"
      type="submit"
    >
      Sign Up
    </button>
  </form>
</div>
`;
